src/typval.c

```diff
@@ -2028,7 +2026,7 @@ tv_equal(
     // Reduce the limit every time running into it. That should work fine for
     // deeply linked structures that are not recursively linked and catch
     // recursiveness quickly.
-    if (!recursive)
+    if (recursive_cnt == 0)
```

Comment:
In the former version,  `recursive=TRUE` appears 5 times. They are all directly/indirectly called by `tv_equal()`, which means `tv_equal()` may call itself indirectly and recursively

- 3 times are `list_equal()`, `dict_equal()` and `object_equal()` called directly in `tv_equal()`:
https://github.com/vim/vim/blob/7b29cc97d6c1450865969d08c5538a40c304593c/src/typval.c#L2059-L2069
https://github.com/vim/vim/blob/7b29cc97d6c1450865969d08c5538a40c304593c/src/typval.c#L2101-L2105

- twice in `func_equal()`. Currently, `func_equal()` is only called in `tv_equal()`.
https://github.com/vim/vim/blob/7b29cc97d6c1450865969d08c5538a40c304593c/src/typval.c#L1991-L2001

`recursive=FALSE` means that the caller of `tv_equal()` is the entry, it is the first time to call `tv_equal()` directly/indirectly.

`recursive_cnt` is a static value in `tv_equal()` and its initial value is 0. We only change the value of `recursive_cnt` in the form of the following example:
https://github.com/vim/vim/blob/7b29cc97d6c1450865969d08c5538a40c304593c/src/typval.c#L2101-L2105

When the recursive call of `tv_equal()` ends, `recursive_cnt` is 0. So I think we can replace the condition `if (!recursive)` with `if(recursive_cnt == 0)`.